src/components/DropdownMenu/DropdownMenu.tsx

```diff
@@ -1,211 +1,195 @@
-
-        return (
-            <React.Fragment>
-                <div
-                    ref={this.anchorRef}
-                    className={b('switcher-wrapper', switcherWrapperClassName)}
-                    onClick={this.onSwitcherClick}
-                >
-                    {switcherButton}
-                </div>
-                <DropdownMenuPopup
-                    open={this.state.popupIsShowed}
-                    anchorRef={this.anchorRef}
-                    items={items}
-                    onMenuItemClick={this.onMenuItemClick}
-                    onClose={this.handleClose}
-                    popupClassName={popupClassName}
-                    placement={popupPlacement}
-                    size={size}
-                    menuProps={menuProps}
-                >
-                    {children}
-                </DropdownMenuPopup>
-            </React.Fragment>
-        );
-    }
-
-    private renderDefaultSwitcher() {
-        const {size, defaultSwitcherClassName, defaultSwitcherProps, icon} = this.props;
-
-        return (
-            <Button
-                view="flat"
-                size={size}
-                {...defaultSwitcherProps}
-                className={b('switcher-button', defaultSwitcherClassName)}
-                disabled={this.props.disabled}
-            >
-                {icon}
-            </Button>
-        );
-    }
-
-    private onMenuItemClick = (
-        event: React.MouseEvent<HTMLElement, MouseEvent>,
-        action: DropdownMenuItemAction<T>,
-    ) => {
-        action(event, this.props.data);
-        this.props.onMenuToggle();
-        this.hidePopup();
-    };
-
-    private onSwitcherClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
-        if (this.props.disabled) {
-            return;
+    /**
+     * Custom content inside the menu popup.
+     */
+    children?: ReactNode;
+};
+
+export const DropdownMenu = <T,>({
+    items = [],
+    size = 'm',
+    icon = <Icon data={DotsIcon} />,
+    onMenuToggle,
+    hideOnScroll = true,
+    data,
+    disabled,
+    switcher,
+    switcherWrapperClassName,
+    defaultSwitcherProps,
+    defaultSwitcherClassName,
+    onSwitcherClick,
+    menuProps,
+    popupClassName,
+    popupPlacement,
+    children,
+}: DropdownMenuProps<T>) => {
+    const [isPopupShown, setPopupShown] = useState(false);
+    const anchorRef = useRef<HTMLDivElement | null>(null);
+
+    // these callbacks are mutually dependent;
+    // for this reason, it's easier to define them in a single `useMemo()`
+    const [handleScroll, showPopup, hidePopup] = useMemo(() => {
+        function handleScroll(event: Event) {
+            if ((event.target as Node).contains(anchorRef.current)) {
+                onMenuToggle?.();
+                hidePopup();
+            }
         }
 
-        this.props.onMenuToggle();
-
-        if (this.props.onSwitcherClick) {
-            this.props.onSwitcherClick(e);
+        function showPopup() {
```

Comment:
It is better to use `useEffect` instead of memoization

```js
const onSwitcherClick = useCallback(() => {
  if (disabled) {
    return;
  }

  onMenuToggle();
  onSwitcherClick?.();

  setPopupShown((isPopupShown) => {
    return !isPopupShown;
  });
}, []);

useEffect(() => {
  if (!isPopupShown || !hideOnScroll) {
    return;
  }

  document.addEventListener('scroll', handleScroll, true);

  return () => {
    document.removeEventListener('scroll', handleScroll, true);
  };
}, [isPopupShown, hideOnScroll]);
```

Reply:
looks better indeed, made an update, thanks